shared/actions/chat.js

```diff
@@ -561,6 +576,31 @@ function * _selectConversation (action: SelectConversation): SagaGenerator<any,
   const inbox = yield select(inboxSelector)
   if (inbox) {
     yield put({type: Constants.updateMetadata, payload: {users: inbox.get('participants').filter(p => !p.you).map(p => p.username).toArray()}})
+    // If we're switching to a badged conversation, unbadge it.
+    const conversationStateSelector = (state: TypedState) => state.chat.get('conversationStates', Map()).get(action.payload.conversationIDKey)
+    const conversationState = yield select(conversationStateSelector)
+    if (conversationState && conversationState.messages !== null) {
+      const conversationID = keyToConversationID(action.payload.conversationIDKey)
+      const msgID = conversationState.messages.get(conversationState.messages.size - 1).messageID
+      yield call(localMarkAsReadLocalRpcPromise, {
+        param: {conversationID, msgID},
+      })
+    }
+  }
+}
+
+function * _changedFocus (action: ChangedFocus): SagaGenerator<any, any> {
+  const appFocused = action.payload
+  // Reselect the current Chat conversation, to give badging a chance to
+  // update based on the focus change.
+  const selectedSelector = (state: TypedState) => state.chat.get('selectedConversation')
+  const conversationIDKey = yield select(selectedSelector)
+  const routeSelector = (state: TypedState) => state.routeTree.get('routeState').get('selected')
+  const selectedTab = yield select(routeSelector)
+  const chatTabSelected = (selectedTab === chatTab)
+
+  if (conversationIDKey && appFocused && chatTabSelected) {
+    yield put(selectConversation(conversationIDKey, true))
```

Comment:
can we just do the badging logic and just pull that out so selecting a conversation or handling a focus changed does this. i'm worried about the side effects of selecting the conversation again (like loading more messages etc)